libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -761,6 +761,33 @@ void GCS_MAVLINK::handle_radio_status(mavlink_message_t *msg, DataFlash_Class &d
     }
 }
 
+/*
+  return a timesync request
+  Sends back ts1 as received, and tc1 is the local timestamp in usec
+ */
+void GCS_MAVLINK::handle_timesync(mavlink_message_t *msg)
+{
+    // Decode incoming timesync message
+    mavlink_timesync_t tsync;
+    mavlink_msg_timesync_decode(msg, &tsync);
+
+    // Create new timestruct to return
+    mavlink_timesync_t rsync;
+    rsync.ts1 = tsync.ts1;
+    rsync.tc1 = AP_HAL::micros64();
+
+    // If message is sent with tc > 1, calculate the offset and return that instead
+    if (tsync.tc1 > 0) {
+        rsync.tc1 = tsync.tc1 - AP_HAL::micros64();
```

Comment:
@meee1 Yes this is correct if using this message to try to sync the absolute time of the flight controller, but my intent was to return the time since boot as that is a more reliable and consistent timestamp and is more useful for correlating log timestamps, EKF etc.  The absolute time changes according to quality or absence of GPS lock.
The two fields of the TIMESYNC message don't allow for the same type of synchronisation as NTP which uses 4 fields.  With the two fields of TIMESYNC the best we can achieve is (roundtrip/2) unless using tc1>0 to estimate the incoming offset and return the offset to the remote end.  I've removed that logic in a subsequent commit to clarify the main purpose of the message response.
The px4 code uses tc1>0 to publish a smoothed offset to the remote timestamp within the flight controller, which is probably not what we want to do.

So thanks for raising this important question :)  Which is, do we want to return the time since boot, or the absolute time?